community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/pipes/ExpandIntoPipe.scala

```diff
@@ -69,14 +67,64 @@ case class ExpandIntoPipe(source: Pipe,
     }
   }
 
+  private def iterateRelationships(row: ExecutionContext, state: QueryState, fromNode: Node, toNode: Node) = {
+
+    //check degree and iterate from the node with smaller degree
+    val fromDegree = state.query.nodeGetDegree(fromNode.getId, dir)
+    val toDegree = state.query.nodeGetDegree(toNode.getId, dir.reverse())
+    val (start, end, relationships) = if (fromDegree < toDegree)
+      (fromNode, toNode, state.query.getRelationshipsForIds(fromNode, dir, types.types(state.query)))
+    else
+      (toNode, fromNode, state.query.getRelationshipsForIds(toNode, dir.reverse(), types.types(state.query)))
+
+    if (toNode == null)
+      Iterator.empty
+    else {
+      new PrefetchingIterator[ExecutionContext] {
+        val connectedRelationships = ArrayBuffer.empty[Relationship]
+
+        override def fetchNextOrNull(): ExecutionContext = {
+          while (relationships.hasNext) {
+            val rel: Relationship = relationships.next()
+            val other = rel.getOtherNode(start)
+            if (end == other) {
+              connectedRelationships.append(rel)
+              return row.newWith2(relName, rel, toName, toNode)
+            }
+          }
+          registerConnection(start, end, connectedRelationships)
```

Comment:
Just curious, as `registerConnection` only happens after the iteration is complete (btw. what happens if this iterator is never exhausted, will they then ever be registered?) 
can it be that we have some race condition here, where the connectedRelationships is not yet registered and then a second call for the two nodes to `internalCreateResults` happens?